sdc_services/models/r2/questionnaire_response.py

```diff
@@ -0,0 +1,64 @@
+from sdc_services.models.r2.observation import Observation
+
+
+class QuestionnaireResponse(object):
+    def __init__(self):
+        self.group = None
+        self.identifier = None
+        self.authored = None
+
+
+    @classmethod
+    def from_json(cls, qnr_json):
+        qnr = cls()
+
+        qnr.group = qnr_json['group']
+        qnr.identifier = qnr_json['identifier']
+        qnr.authored = qnr_json['authored']
+
+        return qnr
+
+
+    def as_fhir(self):
+        fhir_json = {
+            'resourceType': self.__class__.__name__,
+            'group': self.group,
+            'identifier': self.identifier,
+        }
+        # filter out unset attributes
+        filtered_fhir_json = {k:v for k, v in fhir_json.items() if v}
+        return filtered_fhir_json
+
+
+    def walk_answers(self, items=None):
+        """Traverse nested groups and answers, yielding individual answers"""
+
+        if items is None:
+            items = self.group
+
+        for item in items:
+            if item == 'group':
+                yield from self.walk_answers(items['group'])
+            elif item == 'question':
+                for question in items['question']:
+                    for answer in question['answer']:
+                        yield answer
+
+
+    def extract(self):
+        """Extract coded Observations from individual answers"""
+
+        observations = []
+        for answer in self.walk_answers():
+            if 'valueCoding' not in answer:
+                continue
+
+            obs = Observation(
+                # TODO use portal business identifer?
```

Comment:
Regarding the TODO, why would this be needed?  The QNR identifier should be adequate for the source (derivedFrom) but perhaps you're referring to what I think needs to be captured in a CodeableConcept for what the Observation itself is in reference to?

Reply:
I was thinking I would have to use the primary key of the `questionnaire_responses` table, so the portal could easily reference the observations, but I guess those relationships can be setup by the portal at Observation import

Reply:
however you decide is fine - just need an easy route to identify the qnr.id.  keep in mind, we'll need to obtain observation bundles for previous qnr submissions, not just the current, in looking for trigger cases.